src/librustc/hir/lowering.rs

```diff
@@ -4323,53 +4323,147 @@ impl<'a> LoweringContext<'a> {
                 let ohs = P(self.lower_expr(ohs));
                 hir::ExprKind::AddrOf(m, ohs)
             }
-            // More complicated than you might expect because the else branch
-            // might be `if let`.
+            ExprKind::Let(ref pats, ref scrutinee) => {
+                // If we got here, the `let` expression is not allowed.
+                self.sess
+                    .struct_span_err(e.span, "`let` expressions are not supported here")
+                    .note("only supported directly in conditions of `if`- and `while`-expressions")
+                    .note("as well as when nested within `&&` and parenthesis in those conditions")
+                    .emit();
+
+                // For better recovery, we emit:
```

Comment:
This error recovery is not helpful since the errors are not actionable - even if you fix them, `let` is still an unfixable error, so, they mostly create noise.
I'd rather remove this code and just return `hir::ExprKind::Err`.

Reply:
Is there some way I can return `hir::ExprKind::Err` and still avoid `abort_if_errors`? Last I tried this it just produced ICEs because of unused IDs in a HIR validation check.

I agree that it's unfortunate that `let` is an unfixable error, but at least by avoiding `abort_if_errors` we get to 1) continue to typeck and borrowck for other unrelated mistakes the user made, 2) we get to emit the error multiple times if `let` was used mistakenly more than once, 3) because of 2) we now don't need to create a whole lot of files for the `disallowed-positions.rs` test.

cc @estebank

Reply:
>Is there some way I can return hir::ExprKind::Err and still avoid abort_if_errors? Last I tried this it just produced ICEs because of unused IDs in a HIR validation check.

Hm, that's a surprise.
I certainly assumed that `hir::ExprKind::Err` can be returned without aborting.

Reply:
Was a surprise for me also :)

Reply:
If something like
```rust
let _ = self.lower_pats(pats);
let _ = self.lower_expr(scrutinee);
hir::ExprKind::Err
```
doesn't remove the ICEs then this can stay as is.

Reply:
I tried out:
```rust
                let _ = pats.iter().map(|pat| self.lower_pat(pat)).for_each(drop);
                let _ = self.lower_expr(scrutinee);
                hir::ExprKind::Err
```
and got:
```
thread 'rustc' panicked at 'src/librustc/hir/map/hir_id_validator.rs:25:
ItemLocalIds not assigned densely in ::nested_within_if_expr[0]. [...]
```